Heroes Of Pymoli Task

In [ ]:
# Dependencies and Setup

import numpy as np
import pandas as pd

# File to Load
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchases_pd = pd.read_csv(file_to_load)

#show first 5 records of dataframe
#purchases_pd.head()

In [2]:
#Total number of players

grouped_BY_Players_pd = purchases_pd.groupby(["SN"])
print(grouped_BY_Players_pd) 

#grouped_BY_Players_pd.count().head(10)
                                  
#Total_Players = Total_BY_Players.count()
#Total_Players



# Calculate Statistics:

In [3]:
#Total number of players
Player_total = grouped_BY_Players_pd["SN"].sum()
Player_total.count()

576

In [4]:
# Using GroupBy in order to separate the data into fields according to "state" values
grouped_item_pd = purchases_pd.groupby(['Item ID'])

# The object returned is a "GroupBy" object and cannot be viewed normally...
print(grouped_item_pd)



In [5]:
# Total number of unique items
item_total_pd = purchases_pd["Item ID"].value_counts()
unique_item_total = item_total_pd.count()
unique_item_total

183

In [6]:
#Average purchase price

Avg_Price = purchases_pd["Price"].mean()
Avg_Price

3.050987179487176

In [7]:
#Total number of purchases
Total_Purchases = purchases_pd["Purchase ID"].count()
Total_Purchases

780

In [8]:
#calculate total revenue
Total_revenue = purchases_pd["Price"].sum()
Total_revenue

2379.77

In [9]:
# Creating a new DataFrame using stats, but first make a dictionary
summary_table = ({"Number of Unique Items": unique_item_total,
                                    "Average Price": Avg_Price,
                              "Total Purchases": Total_Purchases,
                              "Total Revenue": Total_revenue})
summary_table

{'Number of Unique Items': 183,
 'Average Price': 3.050987179487176,
 'Total Purchases': 780,
 'Total Revenue': 2379.77}

In [10]:
#create dataframe from dictionary
pd.DataFrame.from_dict(summary_table, orient = 'index')

0
Number of Unique Items   183.000000
Average Price              3.050987
Total Purchases          780.000000
Total Revenue           2379.770000

# Gender Analysis

In [11]:
#Gender demographics

#Percentage and Count of Male Players
#Percentage and Count of Female Players
#Percentage and Count of Other / Non-Disclosed

#grouped_gender = grouped_BY_Players_pd["Gender".count


#grouped_gender

#gender_count = grouped_gender_df["Gender"].count()
#gender_count

#gender_count = grouped_gender_df["Gender"].count()
#gender_count


sortforgen_pd = purchases_pd.groupby(['SN','Gender']).first()
print(sortforgen_pd)

sortforgen_pd.count()

#Total_count = gender_df["Gender"].count()
#Total_count



                        Purchase ID  Age  Item ID  \
SN              Gender                              
Adairialis76    Male            467   16      123   
Adastirin33     Female          142   35      175   
Aeda94          Male            388   17      128   
Aela59          Male             28   21      119   
Aelaria33       Male            630   23      171   
Aelastirin39    Male            218   23       76   
Aelidru27       Male            705   22      183   
Aelin32         Male             52   20       60   
Aelly27         Male             43   24      116   
Aellynun67      Male            286   25      153   
Aellyria80      Male            746   23      159   
Aelollo59       Male            203   33        3   
Aenarap34       Male            183   21       15   
Aeral43         Male            263   24       78   
Aeral68         Male            674   43       77   
Aeral97         Male            167   23      157   
Aeralria27      Male             32   10      

Purchase ID    576
Age            576
Item ID        576
Item Name      576
Price          576
dtype: int64

In [12]:
Total_gender = sortforgen_pd.groupby(['Gender']).size().to_frame('Total Count').reset_index()
#Total_gender


In [13]:
#Create dataframe
Gender_sum = Total_gender["Total Count"].sum()
Gender_sum

Total_gender["Percentage of Players"] = Total_gender["Total Count"]/Gender_sum * 100
Total_gender

Gender  Total Count  Percentage of Players
0                 Female           81              14.062500
1                   Male          484              84.027778
2  Other / Non-Disclosed           11               1.909722

In [14]:
#Purchasing Analysis, Gender

#Purchase count
Gender_analysis_pd = purchases_pd.groupby(['Gender'])
print(Gender_analysis_pd)

Purchase_g_count = Gender_analysis_pd["Purchase ID"].count()
#Purchase_g_count


In [15]:
#Average Purchase Price
Average_g_price = Gender_analysis_pd["Price"].mean()
#Average_g_price

In [16]:
#Total Purchase Value
Total_g_price = Gender_analysis_pd["Price"].sum()
#Total_g_price

In [17]:
#Average Purchase Total Per Person By Gender

dfSNgen_pd = purchases_pd.groupby(['SN','Gender'], as_index=False).Price.sum()
#dfSNgen_pd.head()


In [18]:
dfgender_total = dfSNgen_pd.groupby(['Gender']).Price.mean()
#dfgender_total

In [19]:
gendersummary_table_df = ({"Purchase Count": Purchase_g_count,
                                    "Average Price": Average_g_price,
                              "Total Purchase Value": Total_g_price,
                              "Avg Total Purchase Per Person": dfgender_total})
#gendersummary_table_df

In [20]:
#Create dataframe
pd.DataFrame.from_dict(gendersummary_table_df, orient = 'index')

Female         Male  Other / Non-Disclosed
Average Price                    3.203009     3.017853               3.346000
Avg Total Purchase Per Person    4.468395     4.065372               4.562727
Purchase Count                 113.000000   652.000000              15.000000
Total Purchase Value           361.940000  1967.640000              50.190000

# Age Analysis

In [21]:
#The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.)
#Purchase Count
#Average Purchase Price
#Total Purchase Value
#Average Purchase Total per Person by Age Group

In [22]:
# Create the bins in which Data will be held

#bins = [0, 10, 14, 19, 24, 29, 34, 39, 40]
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]
# Create the names for the four bins
group_names = ["Lt 10", "10-14", "15-19", "20-24", "25-29", "29-34", "35-39", "gt 40"]

In [23]:
purchases_pd["Age Categories"] = pd.cut(purchases_pd["Age"], bins, labels=group_names)
#purchases_pd.head()

In [24]:
sortforage_pd = purchases_pd.groupby(['Age Categories'])
print(sortforage_pd)

ByAge_purchasecount = sortforage_pd['Purchase ID'].count()
#ByAge_purchasecount

In [25]:
#Average Purchase Price
Average_A_price = sortforage_pd["Price"].mean()
#Average_A_price

In [26]:
#Total Purchase Value
Total_A_price = sortforage_pd["Price"].sum()
#Total_A_price

In [27]:
#Average Purchase Total per Person by Age Group
dfSNage_pd = purchases_pd.groupby(['SN','Age Categories'], as_index=False).Price.sum()
#dfSNage_pd.head()


In [28]:
dfage_pd = dfSNage_pd.groupby(['Age Categories']).Price.mean()
#dfage_pd

In [29]:
agesummary_table_df = ({"Purchase Count": ByAge_purchasecount,
                                    "Average Price": Average_A_price,
                              "Total Purchase Value": Total_A_price,
                              "Avg Total Purchase Per Person": dfage_pd})
#agesummary_table_df

In [30]:
pd.DataFrame.from_dict(agesummary_table_df, orient = 'index')

Lt 10      10-14       15-19        20-24  \
Average Price                   3.353478   2.956429    3.035956     3.052219   
Avg Total Purchase Per Person   4.537059   3.762727    3.858785     4.318062   
Purchase Count                 23.000000  28.000000  136.000000   365.000000   
Total Purchase Value           77.130000  82.780000  412.890000  1114.060000   

                                    25-29       29-34       35-39      gt 40  
Average Price                    2.900990    2.931507    3.601707   2.941538  
Avg Total Purchase Per Person    3.805195    4.115385    4.763548   3.186667  
Purchase Count                 101.000000   73.000000   41.000000  13.000000  
Total Purchase Value           293.000000  214.000000  147.670000  38.240000

# Various Analyses

In [31]:
#Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
#SN
#Purchase Count
#Average Purchase Price
#Total Purchase Value

In [32]:
#group by SN
dfSN_pd = purchases_pd.groupby(['SN'], as_index=False).Price.sum()
#dfSN_pd.head()

In [33]:
SNlarge_pd = dfSN_pd.nlargest(5, "Price", keep='first')
#SNlarge_pd

In [34]:
Total_Purchases1_pd = purchases_pd.groupby(['SN'], as_index=False).count()
#Total_Purchases1_pd.head()

In [35]:
merge_topspend1_pd = pd.merge(SNlarge_pd, Total_Purchases1_pd, on="SN")
#merge_topspend1_pd

In [36]:
Price_Purchases1_pd = purchases_pd.groupby(['SN'], as_index=False).Price.mean()
#Price_Purchases1_pd.head()

In [37]:
merge_topspend2_pd = pd.merge(merge_topspend1_pd, Price_Purchases1_pd, on="SN")
#merge_topspend2_pd

In [38]:
TopSpenders1_pd = merge_topspend2_pd.drop(columns=['Age', 'Gender', 'Item ID', 'Item Name', 'Price_y', 'Age Categories'])
#TopSpenders1_pd



In [39]:
TopSpenders2_pd = TopSpenders1_pd.rename(columns={"Price_x": "Total Purchase Value", "Purchase ID": "Purchase Count", "Price": "Average Purchase Price"})
TopSpenders2_pd

SN  Total Purchase Value  Purchase Count  Average Purchase Price
0    Lisosia93                 18.96               5                3.792000
1  Idastidru52                 15.45               4                3.862500
2   Chamjask73                 13.83               3                4.610000
3       Iral74                 13.62               4                3.405000
4  Iskadarya95                 13.10               3                4.366667

In [40]:
#Identify the 5 most popular items by purchase count, then list (in a table):
#Item ID
#Item Name
#Purchase Count
#Item Price
#Total Purchase Value

In [45]:
dfSNitem_pd = purchases_pd.groupby(['Item Name', 'Item ID'], as_index=False).count()
#dfSNitem_pd.head()

In [75]:
Purchasecountlarge_pd = dfSNitem_pd.nlargest(5, "Purchase ID", keep='last')
#Purchasecountlarge_pd

In [76]:
dfitempr_pd = purchases_pd.groupby(['Item Name', 'Item ID'], as_index=False).Price.sum()
#dfitempr_pd.head()

In [55]:
#dfprbase_pd = purchases_pd["Price"]
#dfprbase_pd.head()

In [74]:
merge_pop1_pd = pd.merge(dfitempr_pd, Purchasecountlarge_pd, on="Item ID")
#merge_pop1_pd

In [61]:
merge_pop2_pd = merge_pop1_pd.drop(columns=['Item Name_y', 'SN', 'Age', 'Gender', 'Price_y', 'Age Categories'])
merge_pop3_pd = merge_pop2_pd.rename(columns={"Item Name_x": "Item Name", "Purchase ID": "Purchase Count", "Price_x": "Total Purchase Value"})

#merge_pop3_pd

merge_pop3_pd["Item Price"] = merge_pop3_pd["Total Purchase Value"]/merge_pop3_pd["Purchase Count"]
merge_pop3_pd

Item Name  Item ID  \
0     Extraction, Quickblade Of Trembling Hands      108   
1                          Fiery Glass Crusader      145   
2                                       Nirvana       82   
3  Oathbreaker, Last Hope of the Breaking Storm      178   
4                                          Wolf       60   

   Total Purchase Value  Purchase Count  Item Price  
0                 31.77               9        3.53  
1                 41.22               9        4.58  
2                 44.10               9        4.90  
3                 50.76              12        4.23  
4                 28.32               8        3.54

In [69]:
#Identify the 5 most profitable items by total purchase value, then list (in a table):
#Item ID
#Item Name
#Purchase Count
#Item Price
#Total Purchase Value

In [70]:
dfitem_pd = purchases_pd.groupby(['Item Name', 'Item ID'], as_index=False).Price.sum()
#dfitem_pd.head()

In [71]:
itemprofit_pd = dfitem_pd.nlargest(5, "Price", keep='first')
#itemprofit_pd

itemprofit1_pd = itemprofit_pd.rename(columns={"Price": "Total Purchase Value"})
#itemprofit1_pd

In [72]:
merge_prof1_pd = pd.merge(dfSNitem_pd, itemprofit1_pd, on="Item ID")
#merge_prof1_pd

In [80]:
merge_prof2_pd = merge_prof1_pd.drop(columns=['Item Name_y', 'SN', 'Age', 'Gender', 'Price', 'Age Categories', 'Item Name_y'])
merge_prof3_pd = merge_prof2_pd.rename(columns={"Item Name_x": "Item Name", "Purchase ID": "Purchase Count"})
merge_prof3_pd["Item Price"] = merge_prof3_pd["Total Purchase Value"]/merge_prof3_pd["Purchase Count"]
merge_prof3_pd

Item Name  Item ID  Purchase Count  \
0                          Fiery Glass Crusader      145               9   
1                                  Final Critic       92               8   
2                                       Nirvana       82               9   
3  Oathbreaker, Last Hope of the Breaking Storm      178              12   
4                                Singed Scalpel      103               8   

   Total Purchase Value  Item Price  
0                 41.22        4.58  
1                 39.04        4.88  
2                 44.10        4.90  
3                 50.76        4.23  
4                 34.80        4.35

#3 Observable Trends

1) Most purchases are made for ages 20-24 ($1114 for total purchase value) and most customers are male (about 84%).
2) Purchase count wasn't always consistent for total purchase value, but the item with the highest purchase value also had
the highest purchase count (Oathbreaker.)
3) Males spent and purchased more than females. However, the average price (not total) was higher for females ($4.47 vs. $4.07)